I2S (Inter-IC Sound) driver API

mnkp opened the pull request:
I2S (Inter-IC Sound) serial interface can carry relatively large amount of data and needs to do it without any interruptions. The transmission is continuous. The design of this API is based around TX and RX queues which need to be filled / emptied by the user application. In a typical case an application will pre-fill the TX queue with some data and start transmission via a trigger command. This is a similar design to those used by Linux and other embedded OSes.

To prevent unnecessary copying of data buffers between an application and the I2S driver the TX/RX queues are built upon Zephyr memory slabs. A memory block containing data to transmit will be reserved by the application and released by the I2S driver after all data are sent, similarly a memory block that will contain received data will be reserved by the I2S driver and released by the application.

The API allows to configure TX/RX stream format independently. The transmission/reception can start/end independently or synchronously. In case of transmit/reception error the respective queue is blocked to allow application to re-synchronize the data flow.

Both blocking and non-blocking calls are supported.

The different possible scenarios of transmit/receive clock, frame sync and data signal connections between master and slave devices are supposed to be handled by the DTS as signal routing never changes at run time. 

The API is not complete and will have to be extended in the future. Nevertheless a large class of I2S applications is already supported.

This PR includes an example I2S application at samples/drivers/i2s/src/main.c

Jira: ZEP-230
Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp committed:
drivers: Add I2S (Inter-IC Sound) driver API

Origin: Original

Jira: ZEP-230
Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

dbkinder commented on include/i2s.h:
```diff
@@ -0,0 +1,418 @@
+ * Word Select (WS) and Serial Data (SD) signals are sampled on the falling edge
+ * of the clock signal (SCK). The rising edge of the frame sync signal (WS)
+ * indicates the start of the PCM word. The frame sync has an arbitrary length,
+ * however has to fall before the start of the next frame. An arbitrary number
```
some words are missing here?

mnkp replied:
It refers to the frame sync signal. Should the sentence read: "however it has to fall"?

dbkinder replied:
adding "it" works.  Thanks.

dbkinder reviewed (changes requested):
One question, otherwise looks OK.  We should check again once the docs are generated.

tbursztyka commented on include/i2s.h:
```diff
@@ -0,0 +1,418 @@
+	i2s_fmt_t format;
+	i2s_opt_t options;
+	u32_t frame_clk_freq;
+	struct k_mem_slab *mem_slab;
```
Seems that this mem slab could go away of this structure.

mnkp replied:
mem_slab here is a pointer to the pool of TX/RX memory blocks, not a single buffer. That's how I2S driver knows where to release TX memory blocks and where from reserve RX memory blocks. Unlike in networking API memory slabs belong to/are initialized by the application, not the driver.

I'll update inline documentation to provide more information about initializing memory slabs.

tbursztyka commented on include/i2s.h:
```diff
@@ -0,0 +1,418 @@
+#define I2S_OPT_LOOPBACK                    (1 << 7)
+
+
+typedef u32_t i2s_dir_t;
```
same as i2s_fmt_t. Actually, why not an enum as you did for i2s_trigger_cmd?

mnkp replied:
I didn't use enum as I2S_DIR_RX, I2S_DIR_TX may be combined with a bitwise operator, e.g. (I2S_DIR_RX | I2S_DIR_TX) to perform operation on both directions simultaneously. Enum would have to be defined as:
```
enum i2s_dir {
    I2S_DIR_RX,
    I2S_DIR_TX,
    I2S_DIR_RX_TX,
};
```
That's also an option. Should I change it?

As for the typedef let's keep it in line with i2s_opt_t, i2s_fmt_t. If my reasons for using specific types (as stated above) are not good enough I'll remove it from the API.

tbursztyka commented on include/i2s.h:
```diff
@@ -0,0 +1,418 @@
+ */
+
+
+typedef u8_t i2s_fmt_t;
```
I would get rid of such specific typedef, it's just fine to keep u8_t in the config struct and documenting it so it should be populated with the relevant definitions

In device drivers api usually it has been avoided to create such specific types.

mnkp replied:
I've used specific types for two reasons: 
1) it helps user to determine which #define applies to which parameter. 
2) it is likely that i2s_fmt_t and especially i2s_opt_t will have to occupy more than one byte in the future when new options are added. Having a dedicated type, rather than u8_t makes it easier to extend it when the need arises.

tbursztyka commented on include/i2s.h:
```diff
@@ -0,0 +1,418 @@
+#define I2S_FMT_FRAME_CLK_INV               (1 << 5)
+
+
+typedef u8_t i2s_opt_t;
```
same as i2s_fmt_t

tbursztyka reviewed (changes requested):
Where is the driver patch? Would help to evaluate the driver API and config struct.

mnkp commented:
@dbkinder I've generated the docs and they look OK. The only problem is ascii art representing the waveforms. In the rendered HTML they are preceded by a column of "*" characters. That's expected as the waveforms are surrounded by  @verbatim /  @endverbatim doxygen commands. If I remove the "*" the checkpatch complains. The "*" don't look all that terrible but probably should not be there.

mnkp commented on include/i2s.h:
```diff
@@ -0,0 +1,418 @@
+	 * Send all data in the transmit queue and stop the transmission /
+	 * reception. This trigger moves the interface to the READY state.
+	 */
+	I2S_TRIGGER_FLUSH,
```
Woudn't I2S_TRIGGER_DRAIN be a better name?

mnkp commented on include/i2s.h:
```diff
@@ -0,0 +1,418 @@
+ *            | Left channel                  | Right channel                 |
+ * @endverbatim
+ */
+#define I2S_FMT_DATA_FORMAT_I2S             (0 << I2S_FMT_DATA_FORMAT_SHIFT)
```
Format and options #defines are declared as unsigned type, we should probably qualify integer literals as such, i.e.
`(0u << I2S_FMT_DATA_FORMAT_SHIFT)`

mnkp reviewed (commented):
There is one more point I would like to discuss. When working with audio interface people usually talk about I2S. The standard defines both electrical characteristics of the interface as well as data protocol being used. It is limited to stereo data and in practice most often other non-standard protocol extension such as long frame sync is used.

Linux ALSA avoids the name I2S as the name of the interface and calls this part of the library DAI for Digital Audio Interface. Maybe we should do it too?

@tbursztyka I'll need a few more days to upload I2S driver for Atmel SAM SoCs. The driver is working but needs clean up.

dbkinder commented on include/i2s.h:
```diff
@@ -0,0 +1,418 @@
+ * WS changes. Left channel data are sent first indicated by WS = 0, followed
+ * by right channel data indicated by WS = 1.
+ *
+ * @verbatim
```
Instead of using a @verbatim block, you can just indent the text drawing with 4 spaces. This formats the indented lines as a preformatted block and the leading `*` are stripped off.

mnkp replied:
That works well. Thanks!

dbkinder replied:
Thank you for writing API documentation!  :)

lpereira commented:
> Linux ALSA avoids the name I2S as the name of the interface and calls this part of the library DAI for Digital Audio Interface. Maybe we should do it too?

For consistency sake, I'd prefer to leave this as I2S (we have SPI, I2C, etc.).  Makes it more discoverable; without the context from ALSA, I wouldn't know what DAI was referring to.

mnkp committed:
drivers: Add I2S (Inter-IC Sound) driver infrastructure

Only Kconfig/Makefile file for now.

Origin: Original

Jira: ZEP-230
Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp commented:
>> Linux ALSA avoids the name I2S as the name of the interface and calls this part of the library DAI for Digital Audio Interface. Maybe we should do it too?

>For consistency sake, I'd prefer to leave this as I2S (we have SPI, I2C, etc.). Makes it more discoverable; without the context from ALSA, I wouldn't know what DAI was referring to.

That's pretty much what I was thinking. Though we have to be aware that - to some extent - we misuse the name I2S.

mnkp commented:
I've uploaded Atmel SAM I2S driver to my private repository. It's not yet ready to be submitted as PR but the overall driver structure and most of the functionality is already there. It should help evaluate this API:
e96ac0c8e44f804a24595588a096df354a582ad1

One observation that can be made is that the ability for the trigger command to be applied to two directions/queues simultaneously significantly complicates the driver code. E.g it is possible currently to start transmission and reception simultaneously with the command:
```
i2s_trigger(dev_i2s, I2S_DIR_TX | I2S_DIR_RX, I2S_TRIGGER_START);
```
While it may be convenient at times not every application will need it. Ultimately the driver code will always have to be complex even if it's not always required.

I would like to change the description of the trigger command to state that it may be applied to one direction at a time only. This will make it similar to a trigger command in Linux ALSA.

The one drawback is that 
```
i2s_trigger(dev_i2s, I2S_DIR_TX | I2S_DIR_RX, I2S_TRIGGER_FLUSH);
```
which should flush all data from TX queue and then stop transmission and reception will not be supported. If required we would need to come up with a different mechanism to achieve this.

mnkp commented on samples/drivers/i2s/src/main.c:
```diff
@@ -0,0 +1,207 @@
+	k_mem_slab_free(&rx_mem_slab, &rx_block);
+
+	printk("I2S Transmission finished successfully\n");
+}
```
At some point we should have an audio subsystem. I believe though that Zephyr, being an embedded OS, should expose low level drivers like I2S since some users may have special needs / will prefer not to use higher level APIs.

If this sample app is too complex - it seems like it is - I may simplify it. In my opinion we will need another much more sophisticated app to be used as a test.

nashif replied:
@mknp please move this to tests/drivers. We want to keep samples for real world samples that can do useful stuff, i.e. talk to real hardware and use the i2s in this case on a board.

lpereira commented:
> That's pretty much what I was thinking. Though we have to be aware that - to some extent - we misuse the name I2S.

Some people also misuse I2S to make all kinds of weird stuff, such as VHF transmission (FM radios and TV signals are not unheard of). Although, truth be told, people that do this will often find these things regardless of what they're called.

mnkp commented:
I've updated the API:
- i2s_configure and i2s_trigger operate now on one direction (RX or TX) at a time only.
- fixed issue reported by @dbkinder and a few reported by @tbursztyka 
- minor edits to the documentation: extended/cleaned up.
- sample application code refactored to make it simpler.

The Atmel SAM I2S driver corresponding to the new API can be found here: 832ad93ff8023cfab58f9eed853d57d5e76f5f8e

mnkp commented:
I have updated I2S API doxygen documentation. No functional changes were introduced. I've only clarified the API usage: 
- updated description of error codes
- i2s_trigger command usage in regard to the current state of the I2S interface

Since in Zephyr we work with push -f and the previous PR version is lost I've preserved the changes in my local fork. They can be viewed here b38e03622d4eddd005e2b4163a489bbb29c6296f.

All previously requested changes by @nashif, @tbursztyka should be done. Please review.

dbkinder commented on drivers/i2s/Kconfig:
```diff
@@ -0,0 +1,43 @@
+	  Levels are:
+	  0 OFF, do not write
+	  1 ERROR, only write SYS_LOG_ERR
+	  2 WARNING, write SYS_LOG_WRN in adition to previous level
```
s/adition/addition/  here and next few lines

dbkinder reviewed (changes requested):
misspelling in Kconfig, otherwise OK

mnkp reviewed (commented):
@dbkinder the misspelling is fixed. Sorry for such an obvious mistake.

I have two minor questions regarding parameter names otherwise the API looks currently quite good to me.

mnkp commented:
I have removed I2S sample application from this PR. I'll provide an I2S driver test together with Atmel SAM I2S driver.

mnkp commented:
If there are no further issues @dbkinder, @tbursztyka could you please approve this PR. Thanks!

dbkinder reviewed (approved):
+1 for docs

fathomtreelimited commented:
Hi.  Any thoughts about TDM?  This can sometimes be used with I2S or SAI (serial audio interface) peripherals as the stream is similar.

mnkp commented:
Someone would need to make a proposal to extend this I2S API.

One remark: this interface was designed primarily for the I2S/SAI modules, as built into many SoCs. We would need first to extend it to encompass audio codecs, i.e. the other side I2S/SAI modules are communicating with, and then add TDM on top of it. Probably the best approach would be the one used by Linux ALSA, which is adding a couple of functions handling assignment of TDM slots to the data stream.

sathishkuttan commented:
> i2s_configure and i2s_trigger operate now on one direction (RX or TX) at a time only.

I have hardware that supports bidirectional I2S (formats being the same in both directions, of course) and I would like to extend the driver to add support for bidirectional I2S. This also means the API has to be extended, perhaps by adding `I2S_DIR_TXRX` or `I2S_DIR_BOTH` to `enum i2s_dir`.

@mnkp Was the complexity in the particular [Atmel SAM] driver that led to restricting the API to be unidirectional?

> One observation that can be made is that the ability for the trigger command to be applied to two directions/queues simultaneously significantly complicates the driver code

mnkp commented:
@sathishkuttan The current API supports bidirectional I2S transfer, however each direction has to be configured and triggered independently from the other.

The comment referred to the old design of this API, where it was possible to configure/trigger both channels simultaneously.

You may check out the following example tests/drivers/i2s/i2s_api/src/test_i2s_loopback.c to see how to set up simultaneous transfer on both channels in the loopback mode.

sathishkuttan commented:
@mnkp Thanks for the pointer.
Independent triggers for each direction is a good thing. Since the hardware controller is the same and in many cases the registers share bits for RX and TX, it makes the configuration code complex (atleast in the context of my hardware) when configuring TX and RX independently. The configuration code could make better decisions on how to configure if it knows that the application wants to use bidirectional mode.
Would allowing both directions in the configure API be objectionable?

mnkp commented:
As for the decisions I've made when design this API: there exists hardware which allows to configure TX and RX channels independently and software which wants to use this feature so this was one use case the API definitely had to support. However, I agree that in a typical case a user would use the same parameters for TX and RX channels. To make it easier I originally planned to add support for I2S_DIR_BOTH enum but after implementing it it become clear that this increases complexity and code size of the driver with a few darkish corners which were tricky to handle. Dropping this made the driver code leaner. It also doesn't hamper application code, which simply needs to call configuration function twice with the same parameters.

As you wrote, you have a hardware that does not support independent configuration of RX and TX channels. The typical way to handle the scenario where the hardware doesn't support all the features the API is providing is to return an error for unsupported features. E.g. if a user configures TX channel and then tries to configure Rx using incompatible parameters the call should return an error.

sathishkuttan commented:
> The typical way to handle the scenario where the hardware doesn't support all the features the API is providing is to return an error for unsupported features.

I understand this part and it is easy to return error when the hardware/driver does not support the features the API is providing. The problem here, though is, the **API is not providing** the feature that the hardware/driver supports. Specifically, the API does **not** have a way for the application to specify that it wants to configure both TX and RX in the same `configure` call.

I will try to submit a PR to add `I2S_DIR_RXTX` to the enum, the drivers/hardware that need RX and TX to be configured separately can return error and the drivers that **can** configure the RX and TX in the same `configure` call can do so.
